File: frontend_web/src/features/specialist-dashboard/utils/specialistAiReportStructuredSummary.js
function formatJsonSummary(value) {
  if (value == null) {
    return "";
  }
  if (typeof value === "string") {
    return value.trim();
  }
  if (Array.isArray(value)) {
    return value.map(formatJsonSummary).filter(Boolean).join("\n");
  }
  if (typeof value === "object") {
    return Object.entries(value)
      .map(([key, entryValue]) => {
        const formatted = formatJsonSummary(entryValue);
        return formatted ? `${key}: ${formatted}` : "";
      })
      .filter(Boolean)
      .join("\n");
  }
  return String(value).trim();
}

function buildLegacySummaryFallback(raw) {
  if (raw == null) {
    return null;
  }
  if (typeof raw === "string") {
    const trimmed = raw.trim();
    if (!trimmed) {
      return null;
    }
    if (trimmed.startsWith("{") || trimmed.startsWith("[")) {
      try {
        return formatJsonSummary(JSON.parse(trimmed)) || trimmed;
      } catch {
        return trimmed;
      }
    }
    return trimmed;
  }
  if (typeof raw === "object") {
    const formatted = formatJsonSummary(raw);
    return formatted || null;
  }
  const text = String(raw).trim();
  return text || null;
}

function tryParseAiSummaryObject(raw) {
  if (raw == null) {
    return null;
  }
  if (typeof raw === "object" && !Array.isArray(raw)) {
    return raw;
  }
  if (typeof raw !== "string") {
    return null;
  }
  const trimmed = raw.trim();
  if (!trimmed || (!trimmed.startsWith("{") && !trimmed.startsWith("["))) {
    return null;
  }
  try {
    const parsed = JSON.parse(trimmed);
    if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) {
      return null;
    }
    return parsed;
  } catch {
    return null;
  }
}

function readSummaryString(record, key) {
  if (!record || typeof record !== "object") {
    return null;
  }
  const value = record[key];
  if (typeof value === "string" && value.trim()) {
    return value.trim();
  }
  return null;
}

function readSummaryStringArray(record, key) {
  if (!record || typeof record !== "object") {
    return [];
  }
  const value = record[key];
  if (!Array.isArray(value)) {
    return [];
  }
  return value
    .map((item) => (typeof item === "string" ? item.trim() : null))
    .filter(Boolean);
}

export function normalizeAiReportConfidence(value) {
  if (value == null || !Number.isFinite(Number(value))) {
    return null;
  }
  let numeric = Number(value);
  if (numeric > 0 && numeric <= 1) {
    numeric *= 100;
  }
  return Math.round(Math.max(0, Math.min(100, numeric)));
}

export function normalizeAiReportPriority(value) {
  const normalized = String(value || "").trim().toLowerCase();
  if (normalized === "low" || normalized === "medium" || normalized === "high") {
    return normalized;
  }
  return null;
}

const AI_NARRATIVE_SECTION_DEFS = [
  { id: "executive_summary", featured: true },
  { id: "patient_progress_summary" },
  { id: "speech_analysis_summary" },
  { id: "exercise_adherence_summary" },
  { id: "goal_progress_summary" },
];

const AI_LIST_SECTION_DEFS = [
  { id: "clinical_insights", variant: "default" },
  { id: "risks_or_regressions", variant: "warning" },
  { id: "recommendations", variant: "numbered" },
  { id: "next_steps", variant: "default" },
];

/**
 * Parses AI report summary JSON into structured sections for detail rendering.
 * Falls back safely for malformed or legacy summaries.
 */
export function parseAiReportStructuredSummary(raw) {
  const parsed = tryParseAiSummaryObject(raw);
  const emptyFallback = {
    isStructured: false,
    plainTextFallback: null,
    overview: null,
    narrativeSections: [],
    listSections: [],
    contextMetadata: null,
  };

  if (!parsed) {
    return {
      ...emptyFallback,
      plainTextFallback: buildLegacySummaryFallback(raw),
    };
  }

  const narrativeSections = AI_NARRATIVE_SECTION_DEFS
    .map(({ id, featured }) => {
      const content = readSummaryString(parsed, id);
      return content ? { id, content, featured: Boolean(featured) } : null;
    })
    .filter(Boolean);

  const listSections = AI_LIST_SECTION_DEFS
    .map(({ id, variant }) => {
      const items = readSummaryStringArray(parsed, id);
      return items.length > 0 ? { id, items, variant } : null;
    })
    .filter(Boolean);

  const overview = {
    priorityLevel: normalizeAiReportPriority(parsed.priority_level),
    confidencePercent: normalizeAiReportConfidence(parsed.estimated_confidence),
    usedFallback: parsed.used_fallback === true,
    reportType: readSummaryString(parsed, "report_type"),
  };

  const hasStructuredContent = narrativeSections.length > 0
    || listSections.length > 0
    || overview.priorityLevel
    || overview.confidencePercent != null
    || overview.usedFallback;

  if (!hasStructuredContent) {
    return {
      ...emptyFallback,
      plainTextFallback: buildLegacySummaryFallback(raw),
      contextMetadata: parsed.context_metadata ?? null,
    };
  }

  return {
    isStructured: true,
    plainTextFallback: null,
    overview,
    narrativeSections,
    listSections,
    contextMetadata: parsed.context_metadata ?? null,
  };
}

/**
 * Builds AI report detail sections from raw summary JSON.
 * Used by mapAiReportDetail; kept pure for testability.
 */
export function buildAiReportDetailSections(rawSummary, flattenedSummary = null) {
  const aiStructuredSummary = parseAiReportStructuredSummary(rawSummary);
  const fallbackContent = flattenedSummary
    || aiStructuredSummary.plainTextFallback
    || null;

  return {
    aiStructuredSummary,
    sections: aiStructuredSummary.isStructured || !fallbackContent
      ? []
      : [{ title: "AI Summary", content: fallbackContent }],
  };
}

File: mobile_app/lib/features/dashboard/models/specialist_ai_report_structured_summary.dart
import 'dart:convert';

/// Structured AI report summary fields used for review/edit before PDF approval.
class SpecialistAiReportStructuredSummary {
  const SpecialistAiReportStructuredSummary({
    required this.isStructured,
    this.plainTextFallback,
    this.narrativeSections = const [],
    this.listSections = const [],
    this.priorityLevel,
    this.confidencePercent,
  });

  final bool isStructured;
  final String? plainTextFallback;
  final List<SpecialistAiReportNarrativeSection> narrativeSections;
  final List<SpecialistAiReportListSection> listSections;
  final String? priorityLevel;
  final int? confidencePercent;

  static const narrativeFieldIds = <String>[
    'executive_summary',
    'patient_progress_summary',
    'speech_analysis_summary',
    'exercise_adherence_summary',
    'goal_progress_summary',
  ];

  static const listFieldIds = <String>[
    'clinical_insights',
    'risks_or_regressions',
    'recommendations',
    'next_steps',
  ];

  factory SpecialistAiReportStructuredSummary.parse(dynamic raw) {
    final parsed = _tryParseObject(raw);
    if (parsed == null) {
      return SpecialistAiReportStructuredSummary(
        isStructured: false,
        plainTextFallback: _legacyFallback(raw),
      );
    }

    final narrative = <SpecialistAiReportNarrativeSection>[];
    for (final id in narrativeFieldIds) {
      final content = _readString(parsed, id);
      if (content != null) {
        narrative.add(
          SpecialistAiReportNarrativeSection(
            id: id,
            content: content,
            featured: id == 'executive_summary',
          ),
        );
      }
    }

    final lists = <SpecialistAiReportListSection>[];
    for (final id in listFieldIds) {
      final items = _readStringList(parsed, id);
      if (items.isNotEmpty) {
        lists.add(
          SpecialistAiReportListSection(
            id: id,
            items: items,
            variant: id == 'risks_or_regressions'
                ? 'warning'
                : id == 'recommendations'
                    ? 'numbered'
                    : 'default',
          ),
        );
      }
    }

    final priority = _normalizePriority(parsed['priority_level']);
    final confidence = _normalizeConfidence(parsed['estimated_confidence']);
    final usedFallback = parsed['used_fallback'] == true;

    final hasStructured = narrative.isNotEmpty ||
        lists.isNotEmpty ||
        priority != null ||
        confidence != null ||
        usedFallback;

    if (!hasStructured) {
      return SpecialistAiReportStructuredSummary(
        isStructured: false,
        plainTextFallback: _legacyFallback(raw),
      );
    }

    return SpecialistAiReportStructuredSummary(
      isStructured: true,
      narrativeSections: narrative,
      listSections: lists,
      priorityLevel: priority,
      confidencePercent: confidence,
    );
  }

  String narrativeText(String fieldId) {
    for (final section in narrativeSections) {
      if (section.id == fieldId) {
        return section.content;
      }
    }
    if (!isStructured &&
        fieldId == 'executive_summary' &&
        plainTextFallback != null) {
      return plainTextFallback!;
    }
    return '';
  }

  String listText(String fieldId) {
    for (final section in listSections) {
      if (section.id == fieldId) {
        return section.items.join('\n');
      }
    }
    return '';
  }

  Map<String, String> toDraftFormMap() {
    final form = <String, String>{};
    for (final id in narrativeFieldIds) {
      form[id] = narrativeText(id);
    }
    for (final id in listFieldIds) {
      form[id] = listText(id);
    }
    return form;
  }

  static Map<String, dynamic> buildUpdatePayload(Map<String, String> form) {
    final payload = <String, dynamic>{};
    for (final id in narrativeFieldIds) {
      payload[id] = (form[id] ?? '').trim();
    }
    for (final id in listFieldIds) {
      payload[id] = (form[id] ?? '')
          .split(RegExp(r'\r?\n'))
          .map((line) => line.trim())
          .where((line) => line.isNotEmpty)
          .toList();
    }
    return payload;
  }

  static bool hasClinicalContent(Map<String, String> form) {
    final payload = buildUpdatePayload(form);
    for (final id in narrativeFieldIds) {
      if ((payload[id] as String).isNotEmpty) {
        return true;
      }
    }
    for (final id in listFieldIds) {
      if ((payload[id] as List).isNotEmpty) {
        return true;
      }
    }
    return false;
  }

  static Map<String, dynamic>? _tryParseObject(dynamic raw) {
    if (raw == null) {
      return null;
    }
    if (raw is Map) {
      return raw.map((key, value) => MapEntry(key.toString(), value));
    }
    if (raw is! String) {
      return null;
    }
    final trimmed = raw.trim();
    if (trimmed.isEmpty ||
        (!trimmed.startsWith('{') && !trimmed.startsWith('['))) {
      return null;
    }
    try {
      final decoded = jsonDecode(trimmed);
      if (decoded is Map) {
        return decoded.map((key, value) => MapEntry(key.toString(), value));
      }
    } catch (_) {
      return null;
    }
    return null;
  }

  static String? _legacyFallback(dynamic raw) {
    if (raw == null) {
      return null;
    }
    if (raw is String) {
      final trimmed = raw.trim();
      return trimmed.isEmpty ? null : trimmed;
    }
    return raw.toString().trim().isEmpty ? null : raw.toString().trim();
  }

  static String? _readString(Map<String, dynamic> record, String key) {
    final value = record[key];
    if (value is String && value.trim().isNotEmpty) {
      return value.trim();
    }
    return null;
  }

  static List<String> _readStringList(Map<String, dynamic> record, String key) {
    final value = record[key];
    if (value is! List) {
      return const [];
    }
    return value
        .map((item) => item is String ? item.trim() : item?.toString().trim())
        .whereType<String>()
        .where((item) => item.isNotEmpty)
        .toList();
  }

  static String? _normalizePriority(dynamic value) {
    final normalized = value?.toString().trim().toLowerCase();
    if (normalized == 'low' ||
        normalized == 'medium' ||
        normalized == 'high') {
      return normalized;
    }
    return null;
  }

  static int? _normalizeConfidence(dynamic value) {
    if (value == null) {
      return null;
    }
    final numeric = value is num ? value.toDouble() : double.tryParse('$value');
    if (numeric == null || !numeric.isFinite) {
      return null;
    }
    var scaled = numeric;
    if (scaled > 0 && scaled <= 1) {
      scaled *= 100;
    }
    return scaled.clamp(0, 100).round();
  }
}

class SpecialistAiReportNarrativeSection {
  const SpecialistAiReportNarrativeSection({
    required this.id,
    required this.content,
    this.featured = false,
  });

  final String id;
  final String content;
  final bool featured;
}

class SpecialistAiReportListSection {
  const SpecialistAiReportListSection({
    required this.id,
    required this.items,
    this.variant = 'default',
  });

  final String id;
  final List<String> items;
  final String variant;
}
